fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/MembershipIndexManager.java

```diff
@@ -0,0 +1,602 @@
+                    " WHERE mtx.subject_id = m.subject_id" +
+                        " AND mtx.property = m.property" +
+                        " AND mtx.object_id = m.object_id" +
+                        " AND mtx.source_id = m.source_id" +
+                        " AND mtx.operation = :deleteOp" +
+                    ")";
+
+    private static final String DELETE_EXISTING_FOR_SOURCE =
+            "INSERT INTO membership_tx_operations" +
+            " (subject_id, property, object_id, source_id, start_time, end_time, tx_id, operation, force_flag)" +
+            " SELECT subject_id, property, object_id, source_id, start_time, end_time, :txId, :deleteOp, :forceFlag" +
+            " FROM membership m" +
+            " WHERE m.source_id = :sourceId";
+
+    private static final String PURGE_ALL_REFERENCES_MEMBERSHIP =
+            "DELETE from membership" +
+            " where source_id = :targetId" +
+                " OR subject_id = :targetId" +
+                " OR object_id = :targetId";
+
+    private static final String PURGE_ALL_REFERENCES_TRANSACTION =
+            "DELETE from membership_tx_operations" +
+            " WHERE tx_id = :txId" +
+                " AND (source_id = :targetId" +
+                " OR subject_id = :targetId" +
+                " OR object_id = :targetId)";
+
+    private static final String COMMIT_DELETES =
+            "DELETE from membership" +
+            " WHERE EXISTS (" +
+                " SELECT TRUE" +
+                " FROM membership_tx_operations mto" +
+                " WHERE mto.tx_id = :txId" +
+                    " AND mto.operation = :deleteOp" +
+                    " AND mto.force_flag = :forceFlag" +
+                    " AND membership.source_id = mto.source_id" +
+                    " AND membership.subject_id = mto.subject_id" +
+                    " AND membership.property = mto.property" +
+                    " AND membership.object_id = mto.object_id" +
+                " )";
+
+
+    private static final String COMMIT_ENDS =
+            "UPDATE membership m" +
+            " SET end_time = (" +
+                " SELECT mto.end_time" +
+                " FROM membership_tx_operations mto" +
+                " WHERE mto.tx_id = :txId" +
+                    " AND m.source_id = mto.source_id" +
+                    " AND m.subject_id = mto.subject_id" +
+                    " AND m.property = mto.property" +
+                    " AND m.object_id = mto.object_id" +
+                " )" +
+            " WHERE EXISTS (" +
+                "SELECT TRUE" +
+                " FROM membership_tx_operations mto" +
+                " WHERE mto.tx_id = :txId" +
+                    " AND mto.operation = :deleteOp" +
+                    " AND m.source_id = mto.source_id" +
+                    " AND m.subject_id = mto.subject_id" +
+                    " AND m.property = mto.property" +
+                    " AND m.object_id = mto.object_id" +
+                " )";
```

Comment:
The first `WHERE` clause is in order to get the new value for `end_time`, and the `WHERE EXISTS` is to determine which `membership` entries are being ended. But I'll add the operation check to the first subquery, just in case there are multiple operations for membership entry in the current transaction. If I'm remembering correctly, this query ended up kind of dumb looking because h2 doesn't support joins in `UPDATE` queries, so I had to have the two nearly identical subqueries.